src/mam/mod_mam_utils.erl

```diff
@@ -390,6 +391,33 @@ has_chat_marker(Packet) ->
         _                                 -> false
     end.
 
+get_retract_id(Packet) ->
+    case exml_query:subelement_with_ns(Packet, ?NS_FASTEN) of
+        El = #xmlel{name = <<"apply-to">>} ->
+            case exml_query:subelement_with_ns(El, ?NS_RETRACT) of
+                #xmlel{name = <<"retract">>} -> exml_query:attr(El, <<"id">>, none);
+                _ -> none
+            end;
+        _ -> none
+    end.
+
+get_origin_id(Packet) ->
+    exml_query:path(Packet, [{element_with_ns, <<"origin-id">>, ?NS_STANZAID},
+                             {attr, <<"id">>}], none).
+
+tombstone(Packet, OriginID) ->
+    Packet#xmlel{children = [retracted_element(OriginID)]}.
+
+retracted_element(OriginID) ->
+    Timestamp = calendar:system_time_to_rfc3339(erlang:system_time(second), [{offset, "Z"}]),
```

Comment:
I think we need to rework all timestamps in stanzas like this.

Reply:
What exactly do you mean? To use this method instead of the custom code? I think that's a good idea.

Reply:
Yes, exactly! We have some functions in `jlib` that generate timestamps, but since Erlang 21 they can be replaced with the ones from the `calendar` module.